lib/configuration.js

```diff
@@ -15,150 +15,615 @@
+      // project number, set on instance
+      this._projectId = this._givenConfiguration.projectId;
+      return;
+  }
+};
+/**
+ * The _checkLocalProjectNumber function is responsible for determing whether
+ * the _projectNumber property was set by the metadata service and whether or 
+ * not the _projectNumber property should/can be set with a environmental or 
+ * runtime configuration variable. If, upon execution of the 
+ * _checkLocalProjectNumber function, the _projectNumber property has already
+ * been set to a number then it is assumed that this property has been set with
+ * the metadata services response. The metadata value for the project number
+ * always take precedence over any other locally configured project number
+ * value. Given that the metadata service did not set the project number this
+ * function will defer next to the value set in the environment named
+ * `GCLOUD_PROJECT` if it is set and of type string and this string represents a
+ * number (e.g. '123'). If this environmental variable is not set the function
+ * will defer to the _givenConfiguration property if it is of type object and
+ * has a string property named projectId. If none of these conditions are met
+ * then the _projectNumber property will be left at its default value.
+ * @memberof Configuration
+ * @private
+ * @function _checkLocalProjectNumber
+ * @returns {Undefined} - does not return anything
+ */
+Configuration.prototype._checkLocalProjectNumber = function() {
+  if (isString(this._projectNumber)) {
+    // already has been set by the metadata service
+    return;
+  } else if (isString(env.GCLOUD_PROJECT) && !isEmpty(env.GCLOUD_PROJECT)
+    && !isNaN(env.GCLOUD_PROJECT)) {
+      // GCLOUD_PROJECT is set and it is the project number, set on instance
+      this._projectNumber = env.GCLOUD_PROJECT;
+      return;
+  } else if (isPlainObject(this._givenConfiguration)
+    && !isEmpty(this._givenConfiguration.projectId)
+    && !isNaN(this._givenConfiguration.projectId)) {
+      // projectId was set in the runtime configuration and is the project
+      // number, set on instance
+      this._projectNumber = this._givenConfiguration.projectId;
+      return;
+  }
+};
+/**
+ * The _checkLocalServiceContext function is responsible for attempting to
+ * source the _serviceContext objects values from runtime configuration and the
+ * environment. First the function will check the env for known service context
+ * names, if these are not set then it will defer to the _givenConfiguration
+ * property if it is set on the instance. The function will check env variables
+ * `GAE_MODULE_NAME` and `GAE_MODULE_VERSION` for `_serviceContext.service` and
+ * `_serviceContext.version` respectively. If these are not set the 
+ * `_serviceContext` properties will be left at default unless the given runtime
+ * configuration supplies any values as substitutes.
+ * @memberof Configuration
```

Comment:
I don't see JSDoc generation as a high priority issue as this would be internal documentation for a module that is unlikely to grow beyond a small size.